waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/DatabaseMappingUtils.java

```diff
@@ -0,0 +1,128 @@
+
+public class DatabaseMappingUtils {
+
+  public static final String MANUAL_RESOLUTION_TYPE = "MANUAL";
+  public static final String PREFIXED_RESOLUTION_TYPE = "PREFIXED";
+  private static final String INTERRUPTED_MESSAGE = "Interrupted ";
+  private static final String SLOW_METASTORE_MESSAGE = "Metastore was slow to respond so results are omitted";
+  private static final String NULL_POINTER_MESSAGE = "Got nullpointer";
+
+  public static boolean isWhitelisted(String databasePrefix, String database, Map<String, Whitelist> mappedDbByPrefix) {
+    Whitelist whitelist = mappedDbByPrefix.get(databasePrefix);
+    if ((whitelist == null) || whitelist.isEmpty()) {
+      // Accept everything
+      return true;
+    }
+    return whitelist.contains(database);
+  }
+
+  public static List<String> getMappedWhitelistedDatabases(List<String> databases, DatabaseMapping mapping,
+                                                           Map<String, Whitelist> mappedDbByPrefix) {
+    List<String> mappedDatabases = new ArrayList<>();
+    for (String database : databases) {
+      if (DatabaseMappingUtils.isWhitelisted(mapping.getDatabasePrefix(), database, mappedDbByPrefix)) {
+        mappedDatabases.add(mapping.transformOutboundDatabaseName(database));
+      }
+    }
+    return mappedDatabases;
+  }
+
+  public static void shutdownExecutorService(ExecutorService executorService) {
+    executorService.shutdown();
+    try {
+      if (!executorService.awaitTermination(200, TimeUnit.MILLISECONDS)) {
+        executorService.shutdownNow();
+      }
+    } catch (InterruptedException e) {
+      executorService.shutdownNow();
+    }
+  }
+
+  private static List<?> getResultFromFuture(Iterator<DatabaseMapping> iterator, Future<List<?>> future)
+      throws InterruptedException, ExecutionException, TimeoutException {
+    DatabaseMapping mapping = iterator.next();
+    long timeout = mapping.getTimeout();
+    return future.get(timeout, TimeUnit.MILLISECONDS);
+  }
+
+  public static void getDatabasesFromFuture(List<Future<List<?>>> futures, Iterator<DatabaseMapping> iterator,
+                                            List<String> combined, String errorMessage, Logger log) {
+    for (Future<List<?>> future : futures) {
+      try {
+        List<String> result = (List<String>) getResultFromFuture(iterator, future);
+        combined.addAll(result);
+      } catch (InterruptedException e) {
+        log.info(INTERRUPTED_MESSAGE);
+      } catch (ExecutionException e) {
+        log.warn(errorMessage, e.getMessage());
+      } catch (TimeoutException e) {
+        log.warn(SLOW_METASTORE_MESSAGE, e.getMessage());
+      }
+    }
+  }
+
+  public static void getTableMetaFromFuture(List<Future<List<?>>> futures, Iterator<DatabaseMapping> iterator,
+                                            List<TableMeta> combined, String errorMessage, Logger log) {
+    for (Future<List<?>> future : futures) {
+      try {
```

Comment:
This try/catch is repeated three times, you could move it into the `getResultFromFuture` method instead...

Reply:
So I had it at one point like that, but if there is an error, `getResultFromFuture` would have to return null, which then threw a NullPointerException when doing `combined.addAll(null)`.

If I do a null check before I do `addAll`, I can avoid the NPE, but I'm not a fan of returning null from `getResultFromFuture`...